gameswap-service/src/main/java/org/gameswap/web/authentication/JwtAuthFilter.java

```diff
@@ -0,0 +1,100 @@
+import org.gameswap.persistance.UserDAO;
+import org.gameswap.web.authentication.AuthUtils;
+import org.hibernate.CacheMode;
+import org.hibernate.FlushMode;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.context.internal.ManagedSessionContext;
+import org.joda.time.DateTime;
+
+import java.io.IOException;
+import java.security.Principal;
+import java.text.ParseException;
+
+import javax.annotation.Priority;
+import javax.ws.rs.InternalServerErrorException;
+import javax.ws.rs.NotAuthorizedException;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.SecurityContext;
+
+@Priority(1000)
+public class JwtAuthFilter implements ContainerRequestFilter {
+    private UserDAO dao;
+    private SessionFactory sessionFactory;
+
+    public JwtAuthFilter(UserDAO dao, SessionFactory sessionFactory) {
+        this.dao = dao;
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
+        String authHeader = requestContext.getHeaderString(AuthUtils.AUTH_HEADER_KEY);
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            session.setDefaultReadOnly(true);
+            session.setCacheMode(CacheMode.NORMAL);
+            session.setFlushMode(FlushMode.MANUAL);
+            ManagedSessionContext.bind(session);
+            doWork(requestContext, authHeader);
+        }finally{
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    private void doWork(final ContainerRequestContext requestContext, String authHeader) {
+        if (!StringUtils.isBlank(authHeader) && authHeader.split(" ").length == 2) {
+            JWTClaimsSet claimSet;
+            try {
+                claimSet = AuthUtils.decodeToken(authHeader);
+            } catch (ParseException | JOSEException e) {
+                throw new NotAuthorizedException(e);
+            }
+
+            // ensure that the token is not expired
+            if (!new DateTime(claimSet.getExpirationTime()).isBefore(DateTime.now())) {
+                Optional<User> userOptional = dao.find(Long.parseLong(claimSet.getSubject()));
+                Optional<UserPrincipal> transform = userOptional.transform(user -> new UserPrincipal(user.getUsername()));
+                if (transform.isPresent()) {
+                    UserPrincipal userPrincipal = transform.get();
+                    requestContext.setSecurityContext(new SecurityContext() {
+                        @Override
+                        public Principal getUserPrincipal() {
+                            return userPrincipal;
+                        }
+
+                        @Override
+                        public boolean isUserInRole(String s) {
+                            return true;
```

Comment:
userPrincipal just needs to be extended to have roles on it, then we can delegate to it for this check. To start off with the user can access any role annotated resource.